File: lib/android/src/main/java/io/loglens/android/LogLensAndroid.java
package io.loglens.android;

import android.content.Context;
import android.content.pm.ApplicationInfo;

import java.io.File;

import io.loglens.core.FileSink;
import io.loglens.core.LogLens;

/**
 * 안드로이드 부트스트랩 한 줄.
 *
 * <pre>{@code
 * public class App extends Application {
 *     @Override public void onCreate() {
 *         super.onCreate();
 *         LogLensAndroid.install(this);          // logcat 만
 *         // LogLensAndroid.install(this, true); // + 날짜별 파일
 *     }
 * }
 * }</pre>
 *
 * <h2>왜 {@code BuildConfig.DEBUG} 를 안 쓰는가</h2>
 * 라이브러리 모듈의 {@code BuildConfig.DEBUG} 는 <b>그 라이브러리의</b> 빌드타입이다.
 * 앱이 release 로 빌드돼도 라이브러리 쪽이 true 로 남는 조합이 실제로 존재한다.
 * 그러면 릴리스 앱에서 V/D 로그가 통째로 새어나간다. 그래서 앱의
 * {@link ApplicationInfo#FLAG_DEBUGGABLE} 를 런타임에 확인한다 —
 * 이건 항상 <b>앱</b>의 상태다.
 */
public final class LogLensAndroid {

    private LogLensAndroid() {}

    public static void install(Context ctx) {
        install(ctx, false);
    }

    public static void install(Context ctx, boolean withFileSink) {
        LogLens.init(isDebuggable(ctx));
        LogLens.addSink(new LogcatSink());
        if (withFileSink) {
            File dir = new File(ctx.getFilesDir(), "loglens");
            LogLens.addSink(new FileSink(dir, "app", 7));
        }
    }

    /** 앱이 디버그 빌드인가. 라이브러리 빌드타입과 무관하게 항상 정확하다. */
    public static boolean isDebuggable(Context ctx) {
        ApplicationInfo ai = ctx.getApplicationInfo();
        return ai != null && (ai.flags & ApplicationInfo.FLAG_DEBUGGABLE) != 0;
    }
}

File: lib/android/src/main/kotlin/io/loglens/android/LogLensAndroid.kt
package io.loglens.android

import android.content.Context
import android.content.pm.ApplicationInfo
import io.loglens.core.FileSink
import io.loglens.core.LogLens
import java.io.File

/**
 * 안드로이드에서 한 줄로 설정을 끝냅니다.
 *
 * ```
 * class App : Application() {
 *     override fun onCreate() {
 *         super.onCreate()
 *         LogLensAndroid.install(this)          // logcat 만
 *         // LogLensAndroid.install(this, true) // + 날짜별 파일
 *     }
 * }
 * ```
 *
 * ## 왜 BuildConfig.DEBUG 를 안 쓰는가
 *
 * 라이브러리 모듈의 `BuildConfig.DEBUG` 는 **그 라이브러리의** 빌드 타입입니다.
 * 앱을 release 로 빌드해도 라이브러리 쪽이 true 로 남는 조합이 실제로 있습니다.
 * 그러면 릴리스 앱에서 V/D 로그가 전부 새어 나갑니다.
 *
 * 그래서 앱의 [ApplicationInfo.FLAG_DEBUGGABLE] 을 실행 중에 확인합니다.
 * 이건 언제나 **앱**의 상태입니다.
 */
object LogLensAndroid {

    @JvmStatic
    @JvmOverloads
    fun install(ctx: Context, withFileSink: Boolean = false) {
        LogLens.init(isDebuggable(ctx))
        LogLens.addSink(LogcatSink())
        if (withFileSink) {
            LogLens.addSink(FileSink(File(ctx.filesDir, "loglens"), "app", 7))
        }
    }

    /** 앱이 디버그 빌드인지. 라이브러리의 빌드 타입과 무관하게 언제나 정확합니다. */
    @JvmStatic
    fun isDebuggable(ctx: Context): Boolean =
        (ctx.applicationInfo?.flags ?: 0) and ApplicationInfo.FLAG_DEBUGGABLE != 0
}
